pkg/controllers/metrics/state/node.go

```diff
@@ -0,0 +1,250 @@
+	}
+
+	return labels
+}
+
+type nodeScraper struct {
+	cluster  *state.Cluster
+	labelMap map[string]map[*prometheus.GaugeVec][]prometheus.Labels
+}
+
+func newNodeCollector(cluster *state.Cluster) *nodeScraper {
+	return &nodeScraper{
+		cluster:  cluster,
+		labelMap: make(map[string]map[*prometheus.GaugeVec][]prometheus.Labels),
+	}
+}
+
+func (ns *nodeScraper) getName() string {
+	return "nodes"
+}
+
+func (ns *nodeScraper) init(ctx context.Context) {
+	for _, gauge := range []*prometheus.GaugeVec{
+		allocatableGaugeVec,
+		podRequestsGaugeVec,
+		podLimitsGaugeVec,
+		daemonRequestsGaugeVec,
+		daemonLimitsGaugeVec,
+		overheadGaugeVec,
+	} {
+		crmetrics.Registry.MustRegister(gauge)
+	}
+}
+
+func (ns *nodeScraper) update(ctx context.Context) {
+	nodes := make(map[string]struct{})
+	ns.cluster.ForEachNode(func(n *state.Node) bool {
+		if _, ok := ns.labelMap[n.Node.Name]; !ok {
+			logging.FromContext(ctx).Debugf("Tracking new node: %s", n.Node.Name)
+			ns.labelMap[n.Node.Name] = make(map[*prometheus.GaugeVec][]prometheus.Labels)
+		}
+		nodes[n.Node.Name] = struct{}{}
+
+		podRequests := resources.Subtract(n.PodTotalRequests, n.DaemonSetRequested)
+		podLimits := resources.Subtract(n.PodTotalLimits, n.DaemonSetLimits)
+		allocatable := n.Node.Status.Capacity
+		if len(n.Node.Status.Allocatable) > 0 {
+			allocatable = n.Node.Status.Allocatable
+		}
+
+		// Populate  metrics
+		var err error
+		for gaugeVec, resourceList := range map[*prometheus.GaugeVec]v1.ResourceList{
+			overheadGaugeVec:       ns.getSystemOverhead(n.Node),
+			podRequestsGaugeVec:    podRequests,
+			podLimitsGaugeVec:      podLimits,
+			daemonRequestsGaugeVec: n.DaemonSetRequested,
+			daemonLimitsGaugeVec:   n.DaemonSetLimits,
+			allocatableGaugeVec:    allocatable,
+		} {
+			err = multierr.Append(err, ns.set(resourceList, n.Node, gaugeVec))
+		}
+
+		if err != nil {
+			logging.FromContext(ctx).Errorf("Failed to generate gauges for %s: %s", n.Node.Name, err)
+		}
+
+		return true
+	})
+
+	ns.cleanup(ctx, nodes)
+}
+
+func (ns *nodeScraper) reset() {
+	for _, gauge := range []*prometheus.GaugeVec{
+		allocatableGaugeVec,
+		podRequestsGaugeVec,
+		podLimitsGaugeVec,
+		daemonRequestsGaugeVec,
+		daemonLimitsGaugeVec,
+		overheadGaugeVec,
+	} {
+		gauge.Reset()
+	}
+}
+
+func (ns *nodeScraper) cleanup(ctx context.Context, existingNodes map[string]struct{}) {
+	nodesToRemove := []string{}
+	for nodeName := range ns.labelMap {
+		if _, ok := existingNodes[nodeName]; !ok {
+			nodesToRemove = append(nodesToRemove, nodeName)
+		}
+	}
+
+	// Remove all gauges associated with removed node
+	for _, node := range nodesToRemove {
+		gaugeMap := ns.labelMap[node]
+		for gaugeVec, labelSet := range gaugeMap {
+			for _, labels := range labelSet {
+				gaugeVec.Delete(labels)
+			}
+		}
+		delete(ns.labelMap, node)
+	}
+
+	if len(nodesToRemove) > 0 {
+		logging.FromContext(ctx).Debugf("Removed the following node gauges: %s", strings.Join(nodesToRemove, ", "))
```

Comment:
It will print once if one or more nodes are removed since the last scrape. So if nodes are constantly being added and removed it could print, at most, once every 5 seconds (or whatever the configured scrape interval is).

Reply:
Let's remove this log line then. I'm not sure it adds value to users reading the logs.